File: modules/tests/src/commonTest/kotlin/primitives/PassPrimitive.kt
package primitives

import natives.test.MyDictionary
import natives.test.MyEnum
import withLib
import kotlin.test.Test
import kotlin.test.assertTrue

class PassPrimitive {

    @Test
    fun passVoid() = withLib {
        assertTrue(natives.test.passVoid())
    }

    @Test
    fun passChar() = withLib {
        assertTrue(natives.test.passChar('a'))
    }

    @Test
    fun passBoolean() = withLib {
        assertTrue(natives.test.passBoolean(true))
    }

    @Test
    fun passByte() = withLib {
        assertTrue(natives.test.passByte(1.toByte()))
    }

    @Test
    fun passUByte() = withLib {
        assertTrue(natives.test.passUByte(UByte.MAX_VALUE))
    }

    @Test
    fun passShort() = withLib {
        assertTrue(natives.test.passShort(1.toShort()))
    }

    @Test
    fun passUShort() = withLib {
        assertTrue(natives.test.passUShort(UShort.MAX_VALUE))
    }

    @Test
    fun passInt() = withLib {
        assertTrue(natives.test.passInt(99))
    }

    @Test
    fun passUInt() = withLib {
        assertTrue(natives.test.passUInt(UInt.MAX_VALUE))
    }

    @Test
    fun passLong() = withLib {
        assertTrue(natives.test.passLong(9223372036854775805L))
    }

    @Test
    fun passULong() = withLib {
        assertTrue(natives.test.passULong(ULong.MAX_VALUE))
    }

    @Test
    fun passFloat() = withLib {
        assertTrue(natives.test.passFloat(99.9f))
    }

    @Test
    fun passDouble() = withLib {
        assertTrue(natives.test.passDouble(1.1))
    }

    @Test
    fun passString() = withLib {
        assertTrue(natives.test.passString("test string"))
    }

    @Test
    fun passStringN() = withLib {
        assertTrue(natives.test.passStringN(null))
    }

    @Test
    fun passEnum() = withLib {
        assertTrue(natives.test.passEnum(MyEnum.CASE2))
    }

    @Test
    fun passDictionary() = withLib {
        assertTrue(natives.test.passDictionary(MyDictionary(1, 2, 3, 4)))
    }

    @Test
    fun passDictionaryN() = withLib {
        assertTrue(natives.test.passDictionaryN(null))
    }
}

File: modules/tests/natives/testcpp/src/primitives/pass_primitive.cpp
#include <api.hpp>

// Consume

bool pass_void() {
    return true;
}

bool pass_char(uint16_t arg) {
    return arg == 'a';
}

bool pass_boolean(bool arg) {
    return arg == true;
}

bool pass_byte(int8_t arg) {
    return arg == 1;
}

bool pass_ubyte(uint8_t arg) {
    return arg == 255u;
}

bool pass_short(int16_t arg) {
    return arg == 1;
}

bool pass_ushort(uint16_t arg) {
    return arg == 65535u;
}

bool pass_int(int32_t arg) {
    return arg == 99;
}

bool pass_uint(uint32_t arg) {
    return arg == 4294967295u;
}

bool pass_long(int64_t arg) {
    return arg == 9223372036854775805;
}

bool pass_ulong(uint64_t arg) {
    return arg == 18446744073709551615u;
}

bool pass_float(float arg) {
    return arg == 99.9f;
}

bool pass_double(double arg) {
    return arg == 1.1;
}

bool pass_string(KString arg) {
    return std::string(arg.get_data(), arg.get_size()) == "test string";
}

bool pass_string_n(KOptional<KString> arg) {
    return arg.is_none();
}

bool pass_enum(MyEnum arg) {
    return arg == CASE2;
}

bool pass_dictionary(MyDictionary arg) {
    return arg.a == 1 &&
        arg.b == 2 &&
        arg.c == 3 &&
        arg.d == 4;
}

bool pass_dictionary_n(KOptional<MyDictionary> arg) {
    return arg.is_none();
}

bool pass_interface(std::shared_ptr<IMyInterface> arg) {
    return arg->test();
}

bool pass_interface_n(KOptional<std::shared_ptr<IMyInterface>> arg) {
    return arg.is_none();
}
